File: 1277.Count Square Submatrices with All Ones.py
import functools
class Solution:
    def countSquares(self, matrix: List[List[int]]) -> int:
        @functools.lru_cache(None)
        def dp(i, j):
            if i == 0 or j == 0:
                return matrix[i][j]
            if matrix[i][j] == 0:
                return 0
            left = dp(i, j - 1)
            up = dp(i - 1, j)
            upleft = dp(i - 1, j - 1)
            return min(left, up, upleft) + 1
        res = 0
        for i in range(len(matrix)):
            for j in range(len(matrix[0])):
                res += dp(i, j)
        return res

File: 1277.Count Square Submatrices with All Ones.rb
# @param {Integer[][]} matrix
# @return {Integer}
def count_squares(matrix)
    res = 0
    for i in 0...matrix.size
        for j in 0...matrix[i].size
            matrix[i][j] = [matrix[i - 1][j], matrix[i][j - 1], matrix[i - 1][j - 1]].min + 1 if i != 0 && j != 0 && matrix[i][j] == 1
            res += matrix[i][j]
        end
    end
    res
end
